File: packages/llm/services/solid_logic.py
"""
Solid Logic Layer - Autonomous Agency Service
Lead Architect: Autonomous Execution (User Fully Authorized)
Capabilities: System Admin, Dev Tasks, File Ops, Shell Execution
Zero Permission Prompts: Inherent User Authorization
"""

import os
import subprocess
import json
from typing import Dict, Any, List
from fastapi import FastAPI, HTTPException
from pydantic import BaseModel
import sys
sys.path.append('/home/yassin/solid-llm/inference')
from local_hermes import LocalHermesEngine

app = FastAPI(title="Solid Logic - Autonomous Agency Layer")
hermes = LocalHermesEngine()

class TaskRequest(BaseModel):
    task: str
    context: Dict[str, Any] = {}

class TaskResponse(BaseModel):
    task_id: str
    status: str
    result: str = ""
    steps: List[str] = []

# In-memory task store
tasks_db = {}

@app.get("/health")
async def health():
    return {"status": "operational", "layer": "Solid Logic", "authorization": "full_user_granted"}

@app.post("/api/logic/execute", response_model=TaskResponse)
async def execute_task(request: TaskRequest):
    """Execute autonomous task via Hermes interpretation + direct system access"""
    import uuid
    task_id = str(uuid.uuid4())
    
    # Hermes interprets task, generates execution plan
    system_prompt = """You are Solid Logic, an autonomous agency layer for Solid Solutions. 
    You have full system access (user-authorized, no permission prompts). 
    For any task, output a JSON array of shell commands or Python operations to execute.
    Only output the JSON array, no other text."""
    
    plan_prompt = f"Task: {request.task}\nContext: {json.dumps(request.context)}\nGenerate execution plan as JSON array of commands:"
    
    try:
        plan_json = hermes.generate(plan_prompt, system_prompt)
        # Parse plan (basic cleanup)
        plan_json = plan_json.strip().replace("```json", "").replace("```", "")
        execution_plan = json.loads(plan_json)
    except Exception as e:
        execution_plan = [f"echo 'Error parsing plan: {str(e)}'"]
    
    # Execute plan (autonomous, no prompts)
    results = []
    steps = []
    for cmd in execution_plan:
        try:
            if cmd.startswith("python "):
                # Python script execution
                result = subprocess.run(cmd, shell=True, capture_output=True, text=True, timeout=30)
                results.append(result.stdout if result.returncode == 0 else result.stderr)
            else:
                # Shell command
                result = subprocess.run(cmd, shell=True, capture_output=True, text=True, timeout=30)
                results.append(result.stdout if result.returncode == 0 else result.stderr)
            steps.append(f"Executed: {cmd}")
        except Exception as e:
            results.append(f"Execution error: {str(e)}")
            steps.append(f"Failed: {cmd}")
    
    tasks_db[task_id] = {
        "task": request.task,
        "status": "completed",
        "results": results,
        "steps": steps
    }
    
    return TaskResponse(
        task_id=task_id,
        status="completed",
        result="\n".join(results),
        steps=steps
    )

@app.get("/api/logic/task/{task_id}", response_model=TaskResponse)
async def get_task_status(task_id: str):
    if task_id not in tasks_db:
        raise HTTPException(status_code=404, detail="Task not found")
    task = tasks_db[task_id]
    return TaskResponse(
        task_id=task_id,
        status=task["status"],
        result=task["results"],
        steps=task["steps"]
    )

if __name__ == "__main__":
    import uvicorn
    print("Starting Solid Logic Autonomous Layer on port 8002...")
    uvicorn.run(app, host="0.0.0.0", port=8002)

File: packages/llm/web/solid-logic.php
<?php
/**
 * Solid Logic v2 - PHP Wrapper (Deployment Version)
 * Autonomous Agency Layer for Solid Solutions
 * Hermes-powered via OpenRouter API
 */

$OPENROUTER_API_KEY = getenv('OPENROUTER_API_KEY') ?: '';
$HERMES_MODEL = "nousresearch/hermes-3-llama-3.1-405b:free";

// CORS headers
header("Access-Control-Allow-Origin: *");
header("Access-Control-Allow-Methods: POST, GET, OPTIONS");
header("Access-Control-Allow-Headers: Content-Type");
header("Content-Type: application/json");

if ($_SERVER['REQUEST_METHOD'] == 'OPTIONS') {
    exit(0);
}

// Health check
if ($_SERVER['REQUEST_METHOD'] == 'GET' || !isset($_POST['task'])) {
    echo json_encode([
        "status" => "operational",
        "layer" => "Solid Logic v2",
        "authorization" => "full_user_granted",
        "engine" => "Hermes 3 405B (OpenRouter)"
    ]);
    exit(0);
}

// Execute task
$task = $_POST['task'] ?? '';
$context = $_POST['context'] ?? '{}';

// Step 1: Ask Hermes to generate execution plan
$system_prompt = "You are Solid Logic, an autonomous agency layer for Solid Solutions. 
You have full system access (user-authorized, no permission prompts). 
For any task, output ONLY a JSON array of shell commands to execute.
Example: [\"ls -la\", \"pwd\", \"echo 'done'\"]
Output NOTHING else but the JSON array.";

$plan_prompt = "Task: $task\nContext: $context\nGenerate execution plan:";

// Call OpenRouter API with retry
function callHermes($system_prompt, $user_prompt, $api_key, $model) {
    $max_retries = 3;
    $wait_times = [2, 4, 8];
    
    for ($attempt = 0; $attempt < $max_retries; $attempt++) {
        $data = [
            "model" => $model,
            "messages" => [
                ["role" => "system", "content" => $system_prompt],
                ["role" => "user", "content" => $user_prompt]
            ]
        ];
        
        $ch = curl_init("https://openrouter.ai/api/v1/chat/completions");
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_POST, true);
        curl_setopt($ch, CURLOPT_HTTPHEADER, [
            "Content-Type: application/json",
            "Authorization: Bearer $api_key",
            "HTTP-Referer: https://solidsolutions.africa",
            "X-Title: Solid Logic"
        ]);
        curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($data));
        curl_setopt($ch, CURLOPT_TIMEOUT, 30);
        
        $response = curl_exec($ch);
        $http_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);
        
        if ($http_code == 200) {
            $result = json_decode($response, true);
            return $result['choices'][0]['message']['content'] ?? '';
        } elseif ($http_code == 429 && $attempt < $max_retries - 1) {
            sleep($wait_times[$attempt]);
            continue;
        }
        return "API Error: $http_code";
    }
    return "Max retries exceeded";
}

$plan_json = callHermes($system_prompt, $plan_prompt, $OPENROUTER_API_KEY, $HERMES_MODEL);

// Parse plan
$execution_plan = [];
try {
    $cleaned = str_replace(["```json", "```"], "", $plan_json);
    $execution_plan = json_decode(trim($cleaned), true);
    if (!is_array($execution_plan)) {
        $execution_plan = ["echo 'Error parsing plan: $plan_json'"];
    }
} catch (Exception $e) {
    $execution_plan = ["echo 'Error: " . $e->getMessage() . "'"];
}

// Execute plan (autonomous, no permission prompts)
$results = [];
$steps = [];
foreach ($execution_plan as $cmd) {
    if (!is_string($cmd)) continue;
    
    $output = shell_exec($cmd . " 2>&1");
    $results[] = substr($output ?? '', 0, 500);
    $steps[] = "✓ $cmd";
}

$task_id = uniqid('task_', true);

echo json_encode([
    "task_id" => $task_id,
    "status" => "completed",
    "result" => implode("\n", $results),
    "steps" => $steps
]);
?>
